### ISSUE TITLE
macOS build

### ISSUE THREAD

**Issue by thomasza92** [valadaptive/ntsc-rs#36]
Hi,

Great work. Was just wondering if it's possible to get this working on Apple Silicon? I tried to build from source (cargo build) with gstreamer installed via brew but was met with something like 700+ errors.

I wouldn't mind helping to test or patch either!

Thanks

**Comment by valadaptive** [comment]
If you're willing to help with making a macOS build, that'd be great! I haven't been able to do so myself because I don't have any Apple hardware.

What kinds of errors are you getting? Have you installed the gstreamer development headers as well or just the binaries?

**Comment by valadaptive** [comment]
I tried using [FastNoiseLite](https://crates.io/crates/fastnoise-lite) but it was substantially slower--10.2ms/frame vs 6.8ms/frame on my built-in benchmark. You can see my code on the [`noise` branch](https://github.com/valadaptive/ntsc-rs/tree/noise). I did a quick benchmark of the noise-rs version, and it takes 28.459ms/frame.

You can run the benchmark with `RAYON_NUM_THREADS=1 cargo bench --bench filter_profile`. Note that this will give slower results because multithreading is disabled for more consistent performance, but the numbers are consistent relative to each other and that's the important part.

**Comment by valadaptive** [comment]
Try pulling the latest main branch; I've updated it with my own fork of simdnoise. Let me know if it works, and how well the benchmarks run if so.

**Comment by valadaptive** [comment]
I missed some incorrect casts in the original code (they were float-to-int conversions but should've been float-to-int bitcasts) which seem to produce NaNs only on ARM. Try again with the latest main branch, and run the benchmark again--NaNs may have different performance characteristics.

**Comment by valadaptive** [comment]
Since you have an ARM processor, can you try running the benchmark on the main branch then comparing it to the [simd branch](https://github.com/valadaptive/ntsc-rs/tree/simd)? I implemented a SIMD version of the IIR filter and it's faster on x86_64, but I don't have any ARM hardware to test with.

**Comment by sivercone** [comment]
hi, is there instruction how to run this on macos?

**Comment by valadaptive** [comment]
I haven't tried it myself, but [these instructions](https://github.com/valadaptive/ntsc-rs/issues/36#issuecomment-1933440736) should help. Essentially, you'll need to [install GStreamer (use the **development** installer)](https://gstreamer.freedesktop.org/download/#macos), add some variables to your `.zshrc` (use Command+Shift+. to show hidden files), clone this repository, and then `cargo build --target aarch64-apple-darwin --release`.

**Comment by dobkun** [comment]
how to compile on intel mac?

**Comment by valadaptive** [comment]
> how to compile on intel mac?

I think the above instructions *should* work pretty much the same; just replace all instances of `aarch64-apple-darwin` with `x86_64-apple-darwin`.

> Another thing that I think is worth noting from these instructions is how difficult / intimidating it could be for a beginner wishing to use ntsc-rs and needing to compile it from source in order to use it. I am not completely familiar with all the solutions that exist for releasing binaries a[cross](https://github.com/cross-rs/cross) multiple different architectures but I believe it's possible to utilize Github Actions with something like cross.

I definitely want to get something like that set up in the future. There are a few hurdles to overcome:
- I need Apple hardware in order to make sure all the builds actually run properly, and also to get the basic build pipeline functioning before I try to get it running on GH Actions. I've tried running macOS in QEMU but it's not stable.
- I need to figure out how to compile not only the base ntsc-rs application but also the After Effects and OpenFX plugins. As you noted, OpenFX seems to require some sort of plist file. The After Effects plugin is even more complicated; IIRC it needs some [ancient Xcode tools that are still shipped but have no documentation on how to use properly](https://developer.apple.com/forums/thread/675631).
- While it's possible to get standalone unsigned applications to run by going into System Preferences, dynamic libraries like the VFX plugins *must* be signed, which requires a $99/yr Apple developer license.

**Comment by theofabilous** [comment]
I've been able to compile the standalone executable on apple silicon with [homebrew's gstreamer](https://formulae.brew.sh/formula/gstreamer) using apple clang (***not*** homebrew clang, it might work but I haven't tried it). No extra environment variable tinkering was necessary, all I did was clone the repo, ran `brew install gstreamer`, and built the ntsc-rs executable with `cargo build -p gui --release`. Hopefully this helps!

cargo version: `cargo 1.76.0 (c84b36747 2024-01-18)`
homebrew gstreamer version: `1.22.10`
ntsc-rs version: `6cbb5c6` (head of `main` as of writing this)
macos version: Ventura `13.4.1 (22F82)`
cpu: Apple M1 Max

output of `clang --version`:
```
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by Trevva1** [comment]
> # Apple Silicon Instructions

Thanks to everybody for their hard work on this! I had a question for @thomasza92 
Also thanks for your detailed instructions for Mac and I think followed everything up until the end where it says "Navigate to ntsc-rs/target/aarch64-apple-darwin/release and run ntsc-rs-standalone"
Should there be a ntsc-rs-standalone file in that folder directory (an example of how it looks on my end https://imgur.com/m5ou0hC)

I also noticed this appears in my Terminal when I open it - /Users/trevva/.zshrc:12: unmatched "
Is there another step I'm missing? 
I am new to this so apologies if I've missed something basic. TYIA

**Comment by valadaptive** [comment]
Yes, there should be an executable named ntsc-rs-standalone there. Are you getting any errors from `cargo build`?

**Comment by Trevva1** [comment]
Thanks for your quick response @valadaptive 
I double checked over @thomasza92 Compiling notes above and I think the Step 5 (with the cargo build) is where I'm hitting a snag. When I run that in my Mac's terminal I get the error below:
```
trevva@Mac-Studio / % cargo build --target aarch64-apple-darwin --release
error: could not find `Cargo.toml` in `/` or any parent directory
```
I've also supplied an image below of my User directory incase the .zshrc, or other hidden files, are not in their correct place?
https://imgur.com/YZPcjiP

**Comment by valadaptive** [comment]
You need to be cd'd into the repository directory itself.

**Comment by Trevva1** [comment]
Thanks again @valadaptive and apologies for that basic mistake and I think this should be the command below?
```
cd /Users/trevva/ntsc-rs
cargo build --target aarch64-apple-darwin --release
```
However I am getting some errors in the build and maybe I'm still missing a simple fix?
```
Last login: Tue May 21 11:39:46 on ttys000
/Users/trevva/.zshrc:12: unmatched "
trevva@Mac-Studio ~ % cd /Users/trevva/ntsc-rs
trevva@Mac-Studio ntsc-rs % cargo build --target aarch64-apple-darwin --release
   Compiling glib-sys v0.18.1
   Compiling gobject-sys v0.18.0
   Compiling gstreamer-sys v0.21.2
   Compiling gio-sys v0.18.1
   Compiling objc2-encode v2.0.0-pre.2
   Compiling futures-executor v0.3.30
   Compiling parking_lot v0.12.1
   Compiling objc_id v0.1.1
   Compiling futures-channel v0.3.30
   Compiling gstreamer-base-sys v0.21.1
   Compiling simdeez v2.0.0-dev3
   Compiling emath v0.27.0
   Compiling ecolor v0.27.0
   Compiling num-integer v0.1.46
   Compiling unicode-bidi v0.3.15
   Compiling base64 v0.21.7
   Compiling percent-encoding v2.3.1
   Compiling cursor-icon v1.1.0
   Compiling smol_str v0.2.1
   Compiling prettyplease v0.2.16
   Compiling nohash-hasher v0.2.0
   Compiling parking v2.2.0
   Compiling rustix v0.38.31
The following warnings were emitted during compilation:

warning: glib-sys@0.18.1: 

error: failed to run custom build command for `glib-sys v0.18.1`

Caused by:
  process didn't exit successfully: `/Users/trevva/ntsc-rs/target/release/build/glib-sys-0f1ef944fcd0bd30/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=GLIB_2.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=
  pkg-config exited with status code 1
  > PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags glib-2.0 glib-2.0 >= 2.56

  The system library `glib-2.0` required by crate `glib-sys` was not found.
  The file `glib-2.0.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain its parent directory.
  The PKG_CONFIG_PATH environment variable is not set.

  HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `glib-2.0.pc`.

warning: build failed, waiting for other jobs to finish...
The following warnings were emitted during compilation:

warning: gobject-sys@0.18.0: 

error: failed to run custom build command for `gobject-sys v0.18.0`

Caused by:
  process didn't exit successfully: `/Users/trevva/ntsc-rs/target/release/build/gobject-sys-4f4ea8efdc67262c/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=GOBJECT_2.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=
  pkg-config exited with status code 1
  > PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags gobject-2.0 gobject-2.0 >= 2.56

  The system library `gobject-2.0` required by crate `gobject-sys` was not found.
  The file `gobject-2.0.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain its parent directory.
  The PKG_CONFIG_PATH environment variable is not set.

  HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `gobject-2.0.pc`.

The following warnings were emitted during compilation:

warning: gstreamer-sys@0.21.2: 

error: failed to run custom build command for `gstreamer-sys v0.21.2`

Caused by:
  process didn't exit successfully: `/Users/trevva/ntsc-rs/target/release/build/gstreamer-sys-a89bda172224df3d/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=GSTREAMER_1.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=
  pkg-config exited with status code 1
  > PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags gstreamer-1.0 gstreamer-1.0 >= 1.14

  The system library `gstreamer-1.0` required by crate `gstreamer-sys` was not found.
  The file `gstreamer-1.0.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain its parent directory.
  The PKG_CONFIG_PATH environment variable is not set.

  HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `gstreamer-1.0.pc`.

The following warnings were emitted during compilation:

warning: gio-sys@0.18.1: 

error: failed to run custom build command for `gio-sys v0.18.1`

Caused by:
  process didn't exit successfully: `/Users/trevva/ntsc-rs/target/release/build/gio-sys-fe98c5616fd1d160/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=GIO_2.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64-apple-darwin
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_aarch64_apple_darwin
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=
  pkg-config exited with status code 1
  > PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags gio-2.0 gio-2.0 >= 2.56

  The system library `gio-2.0` required by crate `gio-sys` was not found.
  The file `gio-2.0.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain its parent directory.
  The PKG_CONFIG_PATH environment variable is not set.

  HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `gio-2.0.pc`.

trevva@Mac-Studio ntsc-rs % 
```

**Comment by valadaptive** [comment]
Re-read the instructions closelier and make sure you're following the part about .zshrc.

**Comment by Trevva1** [comment]
Hey @valadaptive I've checked and re-applied all of the commands to the .zshrc file as below:
```
# Add Homebrew's executable directory to the front of the PATH
export PATH=/opt/local/bin:$PATH
export CC="/opt/homebrew/Cellar/llvm/18.1.5/bin/clang"
export CXX="/opt/homebrew/Cellar/llvm/18.1.5/bin/clang++"
export CFLAGS="-isysroot $(xcrun -show-sdk-path) ${CFLAGS}"
export CXXFLAGS="-isysroot $(xcrun -show-sdk-path) ${CXXFLAGS}"
export LDFLAGS="-L$(xcrun -show-sdk-path)/usr/lib ${LDFLAGS}"
export SDKROOT=$(xcrun -show-sdk-path)
export PATH="/opt/homebrew/Cellar/llvm/18.1.5/bin/${PATH:+:${PATH}
PKG_CONFIG_PATH="/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/pkgconfig"
PATH="/Library/Frameworks/GStreamer.framework/Versions/1.0/bin:$PATH"
DYLD_FALLBACK_LIBRARY_PATH="/Library/Frameworks/GStreamer.framework/Versions/1.0/lib"
```
When I run a build in the terminal it's still showing `gobject-sys@0.18.0/gstreamer-sys@0.21.2/gio-sys@0.18.1`  failed to run custom build command for `gobject-sys v0.18.0/gstreamer-video-sys v0.21.2` errors

However, it is creating a `libntscrs.rlib` file in the `target/aarch64-apple-darwin/release` folder

Do you have any other suggestions?

**Comment by valadaptive** [comment]
Is it the same error as above, including "The PKG_CONFIG_PATH environment variable is not set."?

**Comment by Trevva1** [comment]
Yes I'm still getting the same errors - apologies for my inexperience and I'll try and do the best I can!

I saw the HINT that appeared in the Terminal and I tried to mess around with the .zshrc file (line 10) below `PKG_CONFIG_PATH="/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/pkgconfig"`
I reduced the line and changed it to `PKG_CONFIG_PATH="/Library/Frameworks/GStreamer.framework"`

New files were created in the `/target/aarch64-apple-darwin/release` folder
The new files were `libopenfx_plugin.dylib` `libntscrs_c_api.a` `libntscrs.rlib` `build-plugin`
I'm not sure if that is helpful at all?

**Comment by valadaptive** [comment]
`PKG_CONFIG_PATH="/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/pkgconfig"` is what I have in my `.zshrc` and it builds fine for me.

Try the following:
- Run `echo $PKG_CONFIG_PATH` in your  terminal to make sure the environment variable is actually being set.
- Install both the **development** and **runtime** GStreamer packages. The instructions only mention the development one, but IIRC the runtime one needs to be installed as well.
- Navigate to `/Library/Frameworks/GStreamer.framework/Versions/1.0/lib/pkgconfig` in Finder to make sure the pkg-config binary actually exists.

**Comment by Trevva1** [comment]
Thanks for those suggestions, but I'm still getting errors.

- I ran the `echo $PKG_CONFIG_PATH` in Terminal, but nothing happened
- The Gstreamer packages have been installed
- It looks like everything is there for the pkg-config - https://imgur.com/adjBrij

The other thing that might be causing issues is that when I open Terminal there is already a note for the last line in .zshrc. Not sure if that is causing issues and if it needs to be resolved?
```
Last login: Wed May 22 13:40:11 on ttys000
/Users/trevva/.zshrc:12: unmatched "
trevva@Mac-Studio
```

**Comment by Trevva1** [comment]
@valadaptive I appreciate your time on this and don't want to take up any more of your time until I've had a thorough look over everything.
Next week when I have time I'll double check over my set up/files and my mac system (as there may be conficts)

**Comment by mobile-bungalow** [comment]
So I haven't read through this entire chain as it's rather long but I would like to point people in the direction of [the new and improved rust AE bindings](https://github.com/AdrianEddy/after-effects), This fork contains build scripts that I have gotten to work on both Windows and MacOs. It uses Lipo to build universal binaries so you only need one MacOs distributable, and it completely handles compiling the PiPl for you. It also contains scripts for writing the plugin bundle and signing it with a development cert! The maintainer has graciously translated tons of examples from adobe, it might be worth a look.

**Comment by valadaptive** [comment]
Hi! Good to hear from you.

In the past couple months, I purchased a Mac Mini and got the standalone application and OpenFX plugin building on macOS--see [the `macos` branch](https://github.com/valadaptive/ntsc-rs/tree/macos) for that work. They're universal binaries so they should work cross-architecture.

All that remains is:
- Building the AE plugin. I have the actual plugin binary building, but haven't figured out the metadata and bundle stuff yet, and I don't have a macOS version of AE to test my work on.
- Signing the app and plugins. My real name unfortunately must be present in the final artifacts because of Apple's developer program, and I have some legal stuff to take care of first.

I might look into the Rust bindings for AE a bit more in the future--I checked them out a bit ago but was worried the build process might be too inflexible (e.g. I might run into an obscure error within their build process and have no way to fix it).

**Comment by valadaptive** [comment]
Hey @thomasza92! I've rewritten the AE plugin in Rust, and I'm hoping you can test it out for me. If you want to test out my other build scripts while you're at it, that would be appreciated (hopefully it should be straightforward).

There are some Cargo "scripts" (really just Rust programs) for building the plugins:
```zsh
cargo xtask macos-bundle --release # for the standalone application
cargo xtask build-ofx-plugin --macos-universal --release # for the OpenFX plugin
cargo xtask macos-ae-plugin --macos-universal --release # for the After Effects plugin.
```

The OpenFX plugin will be under `crates/openfx-plugin/build`, and the others will be under the top-level `build` folder.

They're not signed yet; you should be able to sign them using something [like this snippet](https://github.com/AdrianEddy/after-effects/blob/b5e7454adff1074406770c1c6983bbe5452b941a/AdobePlugin.just#L59) (the plugin path should be `build/ntsc-rs.plugin`).

Thanks so much for helping out with this process!

**Comment by winterwarrior3** [comment]
Howdy! Been following this. I am a Resolve user, would love to text OpenFX Mac version if you need momre testers. Tried to build using the above instructions but was getting errors.

**Comment by valadaptive** [comment]
> Howdy! Been following this. I am a Resolve user, would love to text OpenFX Mac version if you need momre testers. Tried to build using the above instructions but was getting errors.

I've added updated macOS build instructions [in the documentation](https://ntsc.rs/docs/macos-build/). Try following those and let me know if it works (and provide the error messages if it doesn't).

**Comment by valadaptive** [comment]
I've produced a build of ntsc-rs for macOS that runs on my computer (standalone, OpenFX, and After Effects/Premiere). However, I'm not sure if it works on anyone *else's* computer, both because there is no code signing and because I haven't tested it on Apple Silicon.

If you'd like to help me test it out, [here's the latest build](https://drive.proton.me/urls/DYFNPH9C1R#HxGgpqV8Cu0K). Let me know:

- For the **standalone app**:
  - Does it run when you double-click it (with a "this was downloaded from the internet" warning), or does it refuse to run until you manually allow it to in the Privacy & Security settings?
  - Does it start up properly? Give it a few minutes to get past the loading screen.
  - Can you load a variety of image and video formats without errors? If you do run into any errors, screenshot them and post them here.
- For the **OpenFX plugin**:
  - Does it run in DaVinci Resolve downloaded from Blackmagic?
  - Does it run in DaVinci Resolve from the App Store? (I believe Blackmagic says that the App Store version won't load unsigned plugins.)
- For the **After Effects/Premiere plugin**:
  - Does it load (appear in the plugin list) and run properly in After Effects?
  - Does it load (appear in the plugin list) and run properly in Premiere?

I'd like to fill out the full matrix of platforms:

| | Intel | Apple Silicon (ARM) |
|-|-|-|
| Standalone | ❓ | ❓ |
| OpenFX | ❓ | ❓ |
| After Effects | ❓ | ❓ |
| Premiere Pro | ❓ | ❓ |

**Comment by larsjaakko** [comment]
Hello! I randomly stumbled on this thread, quickly tried to run the AE plugin on my M2 macbook pro. Got the error complaining about lack of code signing. 
![Screenshot by Dropbox Capture](https://github.com/valadaptive/ntsc-rs/assets/5289089/3a230dc2-999d-410e-966a-3c13332cd0af)

Might try to build it myself tomorrow, given time, I'll report back in that case.

**Comment by valadaptive** [comment]
@larsjaakko The plugin doesn't "run" per se--are you getting this error when you try to open the .plugin file, or when you open After Effects? You need to install it to `/Library/Application Support/Adobe/Common/Plug-ins/7.0/MediaCore`.

Does it show up in the Stylize category of the effects pane if you do so?

**Comment by larsjaakko** [comment]
@valadaptive Apologies for the vague description — it was installed in the appropriate plugin folder, but the above dialogue pops up immediately on starting After Effects. After clicking it away a few times I was able to open a project and try applying it — so it does indeed appear in the effects pane. Trying to run it does trigger the same dialogue, however, in addition to a few AE specific ones complaining about not being able to run the plugin file.

**Comment by winterwarrior3** [comment]
Reporting in:

The MacOS Standalone seems to be working just fine after some testing. It does prompt an error at first startup due to it not being signed by a developer but that can be bypassed by right clicking on the application and clicking "open". You are then given the opportunity to open regardless of the signature or lackthereof. 

The OpenFX bundle is not being recognized by Resolve so I am unable to test it.

**Comment by valadaptive** [comment]
@larsjaakko If you go into the Privacy & Security settings (in System Settings), is there an option to allow the plugin to run (via the Open Anyway button)? You may need to get the dialog to appear again--apparently the option only shows up for about an hour afterwards.

@winterwarrior3 Ditto with Resolve--do you get a "cannot be opened" dialog when you open Resolve? You may need to re-enable the plugin under the "Video Plug-ins" settings page to get the dialog to appear, and once more after allowing it to run in System Settings. It may also be related to whether you installed the App Store version of Resolve or not.

I'm trying to figure out the best way to release this before I can get a developer certificate--for bureaucratic reasons, it may take a while.

**Comment by winterwarrior3** [comment]
I’ll try that! That may be the issue.

What does it take to get a developers certificate?

**Comment by valadaptive** [comment]
> What does it take to get a developers certificate?

$99/yr, plus your real name on all the software you publish under it.

**Comment by winterwarrior3** [comment]
What’s your best email to contact you? I have an idea.

--

**Comment by valadaptive** [comment]
My username [at] protonmail dot com

**Comment by larsjaakko** [comment]
@valadaptive Success! It worked perfectly, and so far works fine in testing with various videos.

**Comment by winterwarrior3** [comment]
@valadaptive — Emailed you. Also, after removing the bundle and then re-adding it— I was able to get MacOS to prompt me to "Open Anyway" in Privacy & Settings. It works like a charm now after a restart in Resolve. No issues yet!

**Comment by valadaptive** [comment]
@larsjaakko Try installing the debug version of the plugin--let me know if it gives any error messages before crashing:
[ntsc-rs_debug.plugin.zip](https://github.com/user-attachments/files/16191732/ntsc-rs_debug.plugin.zip)

If you could provide a reduced test case .aep for me to reproduce the issue, that'd be great as well. How often does it crash, and is there any discernable pattern to it?

**Comment by valadaptive** [comment]
@larsjaakko Have you had a chance to test out the debug version of the plugin? I want to get the After Effects plugin fixed up before making a new release.

**Comment by valadaptive** [comment]
I think I found the issue with the After Effects plugin. If anyone was having trouble, try installing [the latest version](https://drive.proton.me/urls/DYFNPH9C1R#HxGgpqV8Cu0K) and let me know if it improves things.

**Comment by larsjaakko** [comment]
@valadaptive Apologies for the radio silence! I've been in the middle of moving countries, so notifications took a back seat. 

Just installed the new version, and no crashes to report in about 2 hours of work or so. Previous version crashed several times within a few minutes. Looking very good so far!

**Comment by mobile-bungalow** [comment]
Works over here too!

**Comment by valadaptive** [comment]
I've released official macOS packages in v0.8.0. Running the installers will require some fiddling with Gatekeeper, but once installed, they should work fine with video editing software.

I'm closing this issue since macOS builds have been released now, but let me know if you run into any issues with them.

**Comment by dobkun** [comment]
> Hello! I randomly stumbled on this thread, quickly tried to run the AE plugin on my M2 macbook pro. Got the error complaining about lack of code signing.

go to Settings -> Security & Privacy -> click on the lock -> allow the plugin
